### ISSUE TITLE
Tuplet numbers not always put on the correct side of beamings if <stem> not defined

### ISSUE THREAD

**Issue by fablau** [opensheetmusicdisplay/opensheetmusicdisplay#1449]
Hello Simon and friends.

Here is another issue that should be tackled, and it is about tuplets.

Look at this example below, which includes the <stem> directive to tell the viewer how to orientate the stems (either "up" or "down"):

![stemRight](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/assets/10005316/adde1d6a-f4c4-484c-96c2-b81957d86084)

As you can see, the tuplet numbers are located on the correct side of the group beaming.

Now, look at the same example without the <stem> directive, which should leave the rendering decision of all stem orientations to the OSMD engine:

![stemWrong](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/assets/10005316/00670233-cacf-46c8-b66f-d063bc7e577c)

As you can see, most of the tuplet numbers are located on the wrong side of the beaming despite all stems are correctly automatically oriented. Is that something that can be fixed?

Eager to know your thoughts about it.

Thanks!

**Comment by sschmidTU** [comment]
You've added the same image twice ;)
also, can you share the MusicXML file?

**Comment by fablau** [comment]
The two images are different, look at the tuplet numbers ;)

Here are the XML files, sorry, I forgot about them.

[Archive.zip](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/files/12410358/Archive.zip)

**Comment by sschmidTU** [comment]
Oh, right, the numbers, sorry, for some reason I thought this was about the beams.
Thanks for sample. Will take a look. Probably the numbers should be on the side of the beam always (if not otherwise given in XML).

**Comment by fablau** [comment]
Yes, exactly. They should always be on the beaming side. Thank you Simon!

**Comment by fablau** [comment]
Nice reference! I'd add that that's absolutely true if you have no beamings and, ideally, a bracket. But if you don't have a bracket and have beamings, the number should always be on the beaming side ;)

**Comment by fablau** [comment]
Also, is there a way to tell OSMD to put the tuplet numbers nearer to the beamings when you have this kind of situation (look mostly at the last bar):

![stemDistance](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/assets/10005316/cccdad0d-c1cf-4c4d-916a-77d496123c7a)

The distance between the group and the tuplet number is extremely high.

**Comment by fablau** [comment]
@sschmidTU any thoughts on my last post above? Thanks again.

**Comment by sschmidTU** [comment]
No, there is no way to do that currently. Vexflow does part of the y-placement.

**Comment by fablau** [comment]
Ok, I see... but what about adjusting the side of the tuplet number, as I have explained? Is that something that can be fixed in OSMD?

**Comment by sschmidTU** [comment]
If you mean the first comment / issue opening comment, of course it's possible.

**Comment by fablau** [comment]
Sounds good. Is that something that can be fixed soon/easily?

Also, I am wondering if the second issue that you mentioned being a problem of Vexflow could have been fixed in the latest versions of it? Just wondering... otherwise I'll open an issue over there to have it fixed (or I can see if I can contribute to fix it).

Thanks!

**Comment by sschmidTU** [comment]
It should be relatively easy, but you only see when you do it.

**Comment by fablau** [comment]
Thanks Simon. I just need to know if the first problem (tuplet's number orientation) is something that can be fixed relatively soon (3-4 weeks?) from your team, or if I need to tackle the code myself and do a pull request in case I understand what to do to fix it. The code is very complex, and I might spend entire days to figure out what section is taking care of that.

Thanks again ;)

**Comment by fablau** [comment]
@sschmidTU please, look at this discussion on the Vexflow GitHub issues section:

https://github.com/0xfe/vexflow/issues/1594

What version of Vexflow is OSMD using? Just wondering... thanks.

**Comment by fablau** [comment]
Simon, by looking at OSMD source code it looks like it is still based on Vexflow 1.2.93. Can you confirm? If so, I guess you haven't updated it because of compatibility issues... maybe I can help with that in some way?

**Comment by sschmidTU** [comment]
Yes, it's 1.2.93, as you can see in the package.json.
Vexflow 3+ has layout issues. https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/pull/1139

**Comment by sschmidTU** [comment]
We do apply a VexFlowPatch with lots of improvements, see src/VexFlowPatch/readme

**Comment by fablau** [comment]
I see Simon... I'll review everything and see if we can add a patch or something to handle that sort of tuplet display correctly.

In the meantime, could you please confirm if you or someone on your team can tackle my first issue posted above within the next 3-4 weeks? I am talking about the orientation of the tuplet's numbers to be located on the beaming side when a bracket is not present. I plan to release a new musicXML viewer based on your technology on our website, and it'd be nice to have that issue fixed before its launch. Otherwise, I'll look into it myself.

Thanks again!

**Comment by sschmidTU** [comment]
It's Sunday. You'll get an answer eventually.

**Comment by fablau** [comment]
Of course, and sounds good. ;) Happy Sunday!

**Comment by fablau** [comment]
Can anyone help me here? I think I could fix the patch for tuplets myself and maybe submit a pull request, but I need to know how to "apply" those changes to Vexflow. Do I need to compile those patches aside? I have no idea how to do it. Any changes applied to those JS files are not shown if I run the debugging with "npm start" or if I do a built for distribution with "npm run build"

I am eager to learn more about all this. Thanks.

**Comment by sschmidTU** [comment]
You need to run `npm run prebuildVexflow` to make the changes in VexFlowPatch take effect in `npm start`. This is by the way included in `npm run build` (as part of `prebuild`).

**Comment by sschmidTU** [comment]
By the way, this is also mentioned in the `readme.txt` in the VexFlowPatch folder, where you can also see a list of changes for each vexflow file.
(which should also be expanded for tuplet.js in case of further fixes)

**Comment by fablau** [comment]
Thank you Simon! I read the "readme.txt" but it wasn't clear how to do it. Your explanation above is perfect ;)

Thanks again.

**Comment by fablau** [comment]
Well, I am afraid I posted the above message too early before testing what you suggested... I did this:

`npm run prebuildVexflow`

And then:

`npm start`

But still, I see nothing of what I have done inside those files (the patch JS files) gets executed.
